improve: Extracted original LaTeX when quoting and replying.

userAdityaa opened the pull request:
This commit implemented logic to retrieve the original LaTeX code from .katex-mathml when the user only highlights some part when quote and replying. 
Fixes #31608

Before Change: 
https://github.com/user-attachments/assets/3798272f-a2aa-4df5-8749-48b29fd7774c

After Change: 
https://github.com/user-attachments/assets/f7dfb42e-bdd9-46fc-89dc-fd3e40b25809


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
Please add a test for this.

timabbott commented on web/src/copy_and_paste.ts:
```diff
@@ -433,6 +433,31 @@ export function paste_handler_converter(paste_html: string): string {
+            if (node instanceof Element) {
+                const closestDisplay = node.closest(".katex-display");
+                return closestDisplay !== null;
+            }
```
I feel like the `filter` should check for the `katex-mathml` being actually present inside?

timabbott commented on web/src/copy_and_paste.ts:
```diff
@@ -433,6 +433,31 @@ export function paste_handler_converter(paste_html: string): string {
+                        return `\`\`\`math\n${latexContent}\n\`\`\`\n`;
+                    }
+                }
+            }
```
This seems to only support math blocks, but inline math is also a feature.

userAdityaa replied:
Ok, Solving this issue in a while.

userAdityaa commented:
> Please add a test for this. Also, your images above don't work.

Content is actually in a video format and is working on my end. Could you please verify on your side once more?

timabbott commented on web/tests/copy_and_paste.test.js:
```diff
@@ -55,6 +57,46 @@ run_test("try_stream_topic_syntax_text", () => {
+                <annotation encoding="application/x-tex">$$x^2 + y^2 = z^2$$</annotation>
+            </span>
+        </div>`;
+    assert.equal(copy_and_paste.paste_handler_converter(input), "$$x^2 + y^2 = z^2$$");
```
Is this the actual HTML for inline LaTeX? I would have thought it'd use spans, not divs.

userAdityaa replied:
Yes, Sorry a mistake from my side. It's actually span. 
<img width="565" alt="Screenshot 2024-10-04 at 6 01 55 AM" src="https://github.com/user-attachments/assets/4e6eaa5e-ed02-48ba-92d7-df3513b877fa">

timabbott commented on web/src/copy_and_paste.ts:
```diff
@@ -433,6 +433,36 @@ export function paste_handler_converter(paste_html: string): string {
+        filter(node: Node) {
+            if (node instanceof Element) {
+                const closestDisplay = node.closest(".katex-display");
+                const hasMathML = node.querySelector(".katex-mathml") !== null;
```
Don't use camelCase variable names; that's not our convention.

I'm also not sure why this does the `closest` pattern rather than just processing the `katex` element at the root. For one thing, I think that means it won't work with non-display math inline KaTeX blocks? 

In any case, I pushed a commit with a proposed reworking of this that feels simpler, using the early return pattern and asserts to nest less deeply. It doesn't seem to work; can you debug?

timabbott commented on web/tests/copy_and_paste.test.js:
```diff
@@ -55,6 +57,46 @@ run_test("try_stream_topic_syntax_text", () => {
+    global.Element = window.Element;
+    // Test case for inline LaTeX math
+    let input = `
+        <span class="katex-display">
```
I don't think there's a `katex-display` on inline math.

timabbott commented:
This needs some reworking suggested above; I didn't have time to debug, but pushed a commit to give an idea of the direction I'd suggest for a cleaner implementation of this.

timabbott commented:
Can you post an update on how you've addressed the above feedback?

userAdityaa commented:
Hi @timabbott,
Thank you for the earlier feedback on the LaTeX extraction logic. I’ve made the necessary changes to address the points you raised. Here’s a summary of what I’ve updated:
- Updated variable names to follow the snake_case convention as per project guidelines.
- The use of .closest(".katex-display") as it is necessary to ensure proper handling of display math blocks and compatibility with non-display (inline) KaTeX blocks.
- Improved the logic to process both display and inline KaTeX elements effectively while adhering to code standards.
This update ensures robust handling of all KaTeX blocks and addresses all feedback provided.

shubham-padia commented:
@userAdityaa Your tests are failing, please make sure that tests are passing before you request a re-review. Please also see https://zulip.readthedocs.io/en/latest/contributing/reviewable-prs.html on why videos that need to be downloaded to see the before and after might not be a good idea.
